libass/ass.c

```diff
@@ -186,6 +189,25 @@ static int resize_read_order_bitmap(ASS_Track *track, int max_id)
     return -1;
 }
 
+#define TS_OUT(ev) (ev.Start + ev.Duration)
+
+// This function is not thread safe
+size_t ass_prune_events(ASS_Track *track, size_t threshold, long long deadline)
+{
+    size_t n_size_ini = track->parser_priv->gc_events_size;
+    if (threshold << 10 < track->parser_priv->gc_events_size) {
```

Comment:
Events should be removed uniformly every frame and not periodically in large batches. I've already eliminated similar behavior in cache system and don't want to deal with another. _libass_ is realtime application so latency is more important than throughput.

Reply:
Mhh, I see a few ways to solve this:
- Drop the size threshold: the events would be collected solely based on `deadline`. The user would be responsible to define `lag` appropriately (or `deadline`, if they call `ass_prune_events` directly) to steer the memory footprint.
- Keep the memory footprint threshold, but stop the collection process as soon as $\sum_n SIZE(event_n) < k*gc\textunderscore{}threshold$ is true. $k \in [0; 1.0]$ would be either user-defined, or hardcoded.
- The collection may be aborted whenever $N$ events ($N$ user-defined or hardcoded) have been collected in a pass. This will not protect libass from attacks where more events are added than removed in a frame.